tensorflow_datasets/core/folder_dataset/image_folder.py

```diff
@@ -163,20 +171,33 @@ def _as_dataset(
     if shuffle_files:
       ds = ds.shuffle(len(examples))
     ds = ds.map(_load_example, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+    if not decoders:
+      decode_fn = _default_decode_fn
+    else:
+      decode_fn = functools.partial(
+          self.info.features.decode_example, decoders=decoders)
+
+    ds = ds.map(decode_fn, num_parallel_calls=tf.data.experimental.AUTOTUNE)
```

Comment:
Rather than calling `ds.map` twice. How about moving this logic inside `_load_example` ?

```python
def _load_example(...) -> Dict[str, tf.Tensor]:
  img = tf.io.read_file(path)
  ex = {
      'image': img,
      'label': tf.cast(label, tf.int64),
      'image/filename': path,
  }
  return features.decode_example(ex, decoder=decoder)
```

Reply:
A lambda function seems to be a better way to fix this. The type signature of _load_example becomes bloated with `Callable` and lambda is 1-line of code.